db/write_thread.h

```diff
@@ -136,9 +133,46 @@ class WriteThread {
     Iterator end() const { return Iterator(nullptr, nullptr); }
   };
 
+  struct MultiBatchWriter {
+    std::vector<WriteBatch*> batches;
+    std::atomic<size_t> claimed_cnt;
+    std::atomic<size_t> pending_wb_cnt;
+    ColumnFamilySet* version_set;
+    FlushScheduler* flush_scheduler;
+    bool ignore_missing_column_families;
+    DB *db;
+
+    MultiBatchWriter()
+        : claimed_cnt(0),
+          pending_wb_cnt(0),
+          version_set(nullptr),
+          flush_scheduler(nullptr),
+          ignore_missing_column_families(false),
+          db(nullptr) {}
+
+    explicit MultiBatchWriter(std::vector<WriteBatch*>&& _batch)
+        : batches(_batch),
+          claimed_cnt(0),
+          pending_wb_cnt(_batch.size()),
+          version_set(nullptr),
+          flush_scheduler(nullptr),
+          ignore_missing_column_families(false),
+          db(nullptr) {}
+
+    void SetContext(ColumnFamilySet* _version_set, FlushScheduler* _flush_scheduler,
+                    bool _ignore_missing_column_families, DB* _db) {
+      version_set = _version_set;
+      flush_scheduler = _flush_scheduler;
+      ignore_missing_column_families = _ignore_missing_column_families;
+      db = _db;
+    }
+
+    void ConsumeOne(Writer* writer, size_t claimed);
+    bool ConsumeOne(Writer* writer);
```

Comment:
I prefer to see them implemented in header (they are small and very self-contained). Same goes for functions of `RequestQueue`.

Reply:
Also, I find it a little weird that it requires a `Writer*` input when itself is a member of Writer. How about implementing this in `Writer`?

Reply:
> all and very self-contained). Same goes for functions of `RequestQueue`.


I tried, but met 

```
./db/write_thread.h:171:18: error: ‘WriteBatchInternal’ has not been declared
  171 |       Status s = WriteBatchInternal::InsertInto(
```

As `WriteBatchInternal`  defined in `db/write_batch_internal.h`, which include `db/write_thread.h`, So I think we it cannot be implemented in header.

Reply:
The `RequestQueue` and `WriteThead`  are mutual dependencies, so they cannot be placed in the same header file.